### PR TITLE
Dynamically Change Calendar Date Header Text; closes #2619.

### LINKED ISSUE #2619: Dynamically Change Calendar Date Header Text
TicketCard: currently ticket start and end date header has present tense - sale starts/ sale ends; we need to also cover cases when tickets started or ended in the past ie. `sale started`

see: https://github.com/eventespresso/event-espresso-core/pull/2614#discussion_r398704566

### PR DESCRIPTION
**Pull request by nicolad**
## Problem this Pull Request solves
- Implements `switchTenseForDate` utility function.
- Adds tests for `switchTenseForDate`.
- Use `switchTenseForDate` within `CalendarDateSwitcher`
- Closes: https://github.com/eventespresso/event-espresso-core/issues/2619

## How has this been tested
image

image

### CODE REVIEW

File: assets/src/application/ui/calendars/dateDisplay/CalendarDateSwitcher/index.tsx [hunk 1]
```diff
@@ -1,26 +1,30 @@
...
-	startDate: string;
-}
+const headerFuture = 'sale starts';
+const headerPast = 'sale started';
+const footerFuture = 'sale ends';
+const footerPast = 'sale ended';
 
-const CalendarDateSwitcher: React.FC<Props> = React.memo(
-	({ className, displayDate = DisplayStartOrEndDate.start, footerText, headerText, ...props }) => {
+const CalendarDateSwitcher: React.FC<CalendarDateSwitcherProps> = React.memo(
```
Comment (nicolad) [inline assets/src/application/ui/calendars/dateDisplay/CalendarDateSwitcher/index.tsx]: @tn3rb should we pass ``headerFuture`,  `headerPast`, `footerFuture`, `footerPast`, such that we can control labels in certain situations(maybe) ?

**Comment by nicolad** [comment]
> LGTM 
> May be add tests for uneven cases like empty string, `undefined`, `null`?

I was hesitant to add null checks before because `Datetime` and `Ticket` types has required `startDate` and `endDate`, but added tests, for sanity check.
Thx :+1:

File: assets/src/domain/shared/services/utils/switchTenseForDate/index.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	if (
+		isNil(endDate) ||
+		isNil(footerFuture) ||
+		isNil(footerPast) ||
+		isNil(headerFuture) ||
+		isNil(headerPast) ||
+		isNil(startDate)
+	) {
+		return { footerText: null, headerText: null };
+	}
```
Comment (nicolad) [inline assets/src/domain/shared/services/utils/switchTenseForDate/index.ts]: To avoid confusion or any edge case situations, I've checked each prop, hope that makes sense, if not lemme know.

File: assets/src/domain/shared/services/utils/switchTenseForDate/switchTenseForDate.test.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+			expect(result).toEqual(output);
+		});
+	});
+
+	nullishRestCases.forEach(({ desc, input, output }) => {
+		it(desc, () => {
+			const result = switchTenseForDate(input);
+			expect(result).toEqual(output);
+		});
+	});
```
Comment (manzoorwanijk) [inline assets/src/domain/shared/services/utils/switchTenseForDate/switchTenseForDate.test.ts]: Cool 👍

File: assets/src/domain/shared/services/utils/switchTenseForDate/index.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+const switchTenseForDate = ({
+	endDate,
+	footerFuture,
+	footerPast,
+	headerFuture,
+	headerPast,
+	startDate,
+}: SwitchTenseForDate): { footerText: string; headerText: string } => {
```
Comment (tn3rb) [inline assets/src/domain/shared/services/utils/switchTenseForDate/index.ts]: Ya my apologies because I didn't explain this well at all either...

From that comment in 2614 it was totally clear to me what

```js
const headerText = switchTenseForDate( date, labels.headerPast, labels.headerFuture );
```

meant, but in retrospect I can see that it's not clear at all.

So in my mind `switchTenseForDate()` would ge a generic utility function within the application layer, meaning it would not be specific to any particular use case (domain).
I saw this function being something like:

```js
const switchTenseForDate  = ( date, textForPastDate, textForFutureDate ) => {
	// using the logic you created
	return diff('minutes', parseISO(date), now) < 0 ? textForPastDate: textForFutureDate;
}
```

this could then be used anywhere for any date and for any purpose and would not be specific to the Editor or the `CalendarDateSwitcher` component.

For our current requirements in the `CalendarDateSwitcher` component, we would have to call this utility function for both the start date and end date in order to decide whether to use `footerFuture` or `footerPast` and `headerFuture` or `headerPast` which would be passed into the component via the `CalendarDateLabels` object.

Again, sorry for being "clear as mud" with my previous comments

File: assets/src/domain/shared/services/utils/switchTenseForDate/switchTenseForDate.test.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	},
+];
+
+describe('switchTenseForDate', () => {
+	testCases.forEach(({ desc, input, output }) => {
+		it(desc, () => {
+			const result = switchTenseForDate(input);
+			expect(result).toEqual(output);
+		});
+	});
```
Comment (tn3rb) [inline assets/src/domain/shared/services/utils/switchTenseForDate/switchTenseForDate.test.ts]: i see you like my approach of separating the test case data from the testing logic 👍 

doesn't work for everything but is really great for hammering utilities with any kind of possible input data

File: assets/src/application/services/utilities/date/switchTenseForDate.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import { diff } from '@appServices/utilities';
+import { now } from '@sharedServices/utils/dateAndTime';
+
+const switchTenseForDate = (date: Date, textForPastDate: string, textForFutureDate: string): string | null => {
+	if (isNil(date) || isNil(textForPastDate) || isNil(textForFutureDate)) {
+		return null;
+	}
+
+	return diff('minutes', date, now) < 0 ? textForPastDate : textForFutureDate;
+};
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/src/application/services/utilities/date/switchTenseForDate.ts]: done: https://github.com/eventespresso/event-espresso-core/pull/2629/commits/88bd50e33cf36f45f71b652a4f1b5f35ca9d96a6

File: assets/src/application/ui/calendars/dateDisplay/CalendarDateSwitcher/index.tsx [hunk 7]
```diff
@@ -1,26 +1,19 @@
...
-}
-
-const CalendarDateSwitcher: React.FC<Props> = React.memo(
-	({ className, displayDate = DisplayStartOrEndDate.start, footerText, headerText, ...props }) => {
+const CalendarDateSwitcher: React.FC<CalendarDateSwitcherProps> = React.memo(
+	({ className, displayDate = DisplayStartOrEndDate.start, labels = {}, ...props }) => {
 		const startDate = parseISO(props.startDate) || PLUS_ONE_MONTH;
 		const endDate = parseISO(props.endDate) || PLUS_TWO_MONTHS;
+		const headerText = switchTenseForDate(startDate, labels.headerPast, labels.headerFuture);
+		const footerText = switchTenseForDate(endDate, labels.footerPast, labels.footerFuture);
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/src/application/ui/calendars/dateDisplay/CalendarDateSwitcher/index.tsx]: thx, took your suggestion and went further and applied nullish coalescing operator:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

[the remark this answers is not in the document]
Comment (nicolad) [reply assets/src/application/ui/calendars/dateDisplay/CalendarDateSwitcher/index.tsx]: interesting fact I found from MDN on why this operator is better than OR logical operator:

```
let count = 0;
let text = "";

let qty = count || 42;
let message = text || "hi!";
console.log(qty);     // 42 and not 0
console.log(message); // "hi!" and not ""
```

[the remark this answers is not in the document]
Comment (nicolad) [reply assets/src/application/ui/calendars/dateDisplay/CalendarDateSwitcher/index.tsx]: `??` operator is in TS since v3.7: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#nullish-coalescing

File: assets/src/application/ui/calendars/dateDisplay/CalendarDateSwitcher/types.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { DisplayStartOrEndDate } from '@sharedServices/filterState';
+
+interface CalendarDateLabels {
+	headerPast: string; // sale starts
+	headerFuture: string; // sale started
+	footerPast: string; // sale ends
+	footerFuture: string; // sale ended
+}
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/src/application/ui/calendars/dateDisplay/CalendarDateSwitcher/types.ts]: done, thx
